### PR TITLE
Fix a bug preventing window::raw::{enum_by/enum_by_until} from working after an OS-error is set

### PR DESCRIPTION
**Pull request by ssaavedra**
In case a former call to the Windows API created an error condition (i.e., setting an error on SetErrorEx()), then enum_by and enum_by_until would incorrectly always return an error as a result due to the former error not being cleaned. .Since the contract on safe Rust functions is usually to return their own error (and not carry other former errors to the user) it makes sense that enum_by
would clean the former error with SetErrorEx(0, 0) before operating.

This commit contributes the fix and also a test proving the failure to guard against a regression.

I also change the timings on timer.rs because the test was failing on my machine due to a race condition.

### CODE REVIEW

**Comment by DoumanAsh** [comment]
> Since the contract on safe Rust functions is usually to return their own error (and not carry other former errors to the user) it makes sense that enum_by
would clean the former error with SetErrorEx(0, 0) before operating.

I'm not sure about contract thingy, but these functions are intended to be light wrapper over winapi.
It should do only what it promises (i.e. perform `EnumChildWindows`) via safe interface.
And otherwise behaves like regular winapi calls

It should be user responsibility to actually handle error when it happens.
What you suggests should be done only by high level functions that would invoke `enum_by_until` and friends.

I wouldn't call it a bug, but this behavior could be documented to clarify it and let user know that he may need to clean error if there were any (I didn't look into code for a while now, but I think most functions return io error which would get last error from winapi?)

**Comment by ssaavedra** [comment]
Maybe we disagree on their purpose, but your code fails to execute if for any other former error there was en error set from former code.
If this wrapper is relying on last_os_error() blindly it will fail on many valid use cases.

For example, if you try to enumerate all processes, this will give PermissionDenied errors for some of them; and then for any of those that you have permissions, try to get a window. With your code, it will always return a failure.

If the purpose of the "safe" API is to avoid using unsafe Rust from the raw:: API, the current implementation does not allow to use only safe Rust for this purpose. This is because since any of your safe API calls can return an error, which will SetLastErrorEx, after such an error, there is no "safe API" to reset SetLastErrorEx/last_os_error, forcing your library's user to do so. If they are going to go that far, why use the library at all?

Also, given [the spec](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-enumwindows), it is the EnumWindows' callback responsability to properly use SetLastErrorEx if they want to signal an error, so I'd think it would make sense to encapsulate this low-level behavior, but it's your call.

That's why since this library is trusting the error output of EnumWindows to signal whether there has been an error or not, I believe it should also do a SetLastErrorEx(0, 0) before the EnumWindows call to ensure it is witnessing an actual error. If you wish to not disrupt external users' reliance on last_os_error, you could save it temporarily and then restore it with SetLastErrorEx again before returning Err/Ok.

**Comment by ssaavedra** [comment]
It's not that I really care that much, and it's unlikely I'll write more code that accesses WinAPI from Rust, but for the thing I did, I found this to be really unexpected. 🙂

**Comment by DoumanAsh** [comment]
I just checked your case.
```rust
use winapi::um::winuser::SetLastErrorEx;

#[test]
fn check_enum_by_with_last_error() {
    unsafe {
        SetLastErrorEx(1, 0)
    }

    let result = windows_win::raw::window::enum_by(None, |_| {
        ();
    });

    assert!(result.is_ok());
}
```

It passes for me on current master
So I need to understand how your change can help?

The only place it can incorrectly fail is following:
```rust
#[test]
fn check_enum_by_with_last_error() {
    unsafe {
        SetLastErrorEx(1, 0)
    }

    let result = windows_win::raw::window::enum_by_until(None, |_| {
        1
    });

    assert!(result.is_ok());
}
```

**Comment by DoumanAsh** [comment]
tl;dr
Your idea makes sense for `enum_by_until`

**Comment by DoumanAsh** [comment]
@ssaavedra
I updated your branch.
I'll merge it and release new version today

**Comment by DoumanAsh** [comment]
@ssaavedra `2.4.1` has been released with your fix

**Comment by ssaavedra** [comment]
Thank you! Sorry I was away for a couple of days and didn't see your notifications :-)
